PWGCF/Femto3D/Tasks/femto3dPairTask.cxx

```diff
@@ -0,0 +1,305 @@
+  Filter itsNClsFilter = o2::aod::singletrackselector::itsNCls >= (uint8_t)_itsNCls && o2::aod::singletrackselector::itsChi2NCl < _itsChi2NCl;
+
+  Filter vertexFilter = nabs(o2::aod::singletrackselector::posZ) < _vertexZ;
+
+  Preslice<FilteredTracks> perCollId = o2::aod::singletrackselector::singleCollSelId;
+
+  void init(o2::framework::InitContext&)
+  {
+    IsIdentical = (_sign_1 * _particlePDG_1 == _sign_2 * _particlePDG_2);
+
+    TPCcuts_1 = std::make_pair(_particlePDG_1, _tpcNSigma_1);
+    TOFcuts_1 = std::make_pair(_particlePDG_1, _tofNSigma_1);
+    TPCcuts_2 = std::make_pair(_particlePDG_2, _tpcNSigma_2);
+    TOFcuts_2 = std::make_pair(_particlePDG_2, _tofNSigma_2);
+
+    registry.add("SE", "SE", kTH1F, {{500, 0.005, 5.005, "k*"}});
+    registry.add("ME", "ME", kTH1F, {{500, 0.005, 5.005, "k*"}});
+    registry.add("p_first", "p", kTH1F, {{100, 0., 5., "p"}});
+    registry.add("nsigmaTOF_first", Form("nsigmaTOF_%i", (int)_particlePDG_1), kTH2F, {{100, 0., 5.}, {100, -10., 10.}});
+    registry.add("nsigmaTPC_first", Form("nsigmaTPC_%i", (int)_particlePDG_1), kTH2F, {{100, 0., 5.}, {100, -10., 10.}});
+    if (!IsIdentical) {
+      registry.add("p_second", Form("p_%i", (int)_particlePDG_2), kTH1F, {{100, 0., 5., "p"}});
+      registry.add("nsigmaTOF_second", Form("nsigmaTOF_%i", (int)_particlePDG_2), kTH2F, {{100, 0., 5.}, {100, -10., 10.}});
+      registry.add("nsigmaTPC_second", Form("nsigmaTPC_%i", (int)_particlePDG_2), kTH2F, {{100, 0., 5.}, {100, -10., 10.}});
+    }
+  }
+
+  void process(soa::Filtered<FilteredCollisions> const& collisions, soa::Filtered<FilteredTracks> const& tracks)
+  {
+    int selected_ev_counter = 0;
+
+    if (_particlePDG_1 == 0 || _particlePDG_2 == 0)
+      LOGF(fatal, "One of passed PDG is 0!!!");
+
+    for (auto& collision : collisions) {
+      FemtoParticle* Particle = NULL;
+
+      for (auto& track : tracks.sliceBy(perCollId, collision.index())) {
+
+        if (track.sign() == _sign_1 && (track.p() < _PIDtrshld_1 ? TPCselection(track, TPCcuts_1) : TOFselection(track, TOFcuts_1))) { // filling the map: eventID <-> selected particles1
+          Particle = new FemtoParticle(track.energy(particle_mass(_particlePDG_1)), track.px(), track.py(), track.pz(), track.eta(), track.phi());
+          Particle->SetSign(_sign_1);
+          Particle->SetMagField(collision.magField());
+          selectedtracks_1[track.singleCollSelId()].push_back(Particle);
+
+          registry.fill(HIST("p_first"), track.p());
+          if (_particlePDG_1 == 2212) {
+            registry.fill(HIST("nsigmaTOF_first"), track.p(), track.tofNSigmaPr());
+            registry.fill(HIST("nsigmaTPC_first"), track.p(), track.tpcNSigmaPr());
+          }
+          if (_particlePDG_1 == 1000010020) {
+            registry.fill(HIST("nsigmaTOF_first"), track.p(), track.tofNSigmaDe());
+            registry.fill(HIST("nsigmaTPC_first"), track.p(), track.tpcNSigmaDe());
+          }
```

Comment:
You are replicating in memory everything that you get from the tables with a much more inefficient way of accessing to the stored information. 
This will have a huge penalty in execution time and in memory requirements
This goes against the O2 execution model (SoA vs AoS)
It seems that one of the reasons for storing the information is for extracting the mixed event results
You have to use the mixed events framework that is standard in O2Physics
It works without any kind of data replication

Reply:
Maybe it doesn't seem right but I'll try to explain why I think it's more convenient to use it like this. The main reason is the mixing (I put a more detailed comment about the mixing in the main reply). Since we need to store some info per particle for the mixing later I decided to use this "FemtoParticle" object that contains the bare minimum: pT, Eta, Phi, magField and signedPDG (updated in the last commit wrt to the previous one -- now consumes less memory).

Actually I was trying to store iterators/pointers to the tracks but the ones that are created after slicing or filtering are transient (temporary) ones and cannot be used to access an object (for the mixing) after the selection. So, I don't see any other way to do it apart from creating another additional table which would lead to duplication anyway. But using the class "FemtoPatricle" is more convenient because I can put there needed functions, pass the object to the "FemtoPair" etc.